src/sys/windows/selector.rs

```diff
@@ -1,596 +1,543 @@
-
-        // Poll custom event queue
-        self.readiness_queue.poll(events);
-
-        // Return number of polled events
-        Ok(ret)
+    fn cancel(&mut self) -> io::Result<()> {
+        assert!(self.poll_status == SockPollStatus::Pending);
+        self.afd.cancel(&mut self.iosb.0)?;
+        self.poll_status = SockPollStatus::Cancelled;
+        self.pending_evts = 0;
+        Ok(())
     }
 
-    pub fn select2(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
-        trace!("select; timeout={:?}", timeout);
+    fn mark_delete(&mut self) {
+        if !self.delete_pending {
+            if self.poll_status == SockPollStatus::Pending {
+                drop(self.cancel());
+            }
 
-        // Clear out the previous list of I/O events and get some more!
-        events.clear();
+            self.delete_pending = true;
+        }
+    }
 
-        trace!("polling IOCP");
-        let n = match self.inner.port.get_many(&mut events.statuses, timeout) {
-            Ok(statuses) => statuses.len(),
-            Err(ref e) if e.raw_os_error() == Some(WAIT_TIMEOUT as i32) => 0,
-            Err(e) => return Err(e),
-        };
+    fn feed_event(&mut self) -> Option<Event> {
+        let mut afd_events = 0;
+        self.poll_status = SockPollStatus::Idle;
+        self.pending_evts = 0;
 
-        for status in events.statuses[..n].iter() {
-            // This should only ever happen from the waker.
-            if status.overlapped() as usize == 0 {
-                let token = Token(status.token());
-                if token == WAKE {
-                    continue;
-                }
-                events.events.push(Event::new(Ready::READABLE, token));
-                continue;
+        unsafe {
+            if self.delete_pending {
+                return None;
+            } else if self.iosb.0.u.Status == STATUS_CANCELLED {
+                /* The poll request was cancelled by CancelIoEx. */
+            } else if !NT_SUCCESS(self.iosb.0.u.Status) {
+                /* The overlapped request itself failed in an unexpected way. */
+                afd_events = AFD_POLL_CONNECT_FAIL;
+            } else if self.poll_info.number_of_handles < 1 {
+                /* This poll operation succeeded but didn't report any socket events. */
+            } else if self.poll_info.handles[0].events & AFD_POLL_LOCAL_CLOSE != 0 {
+                self.mark_delete();
+                return None;
+            } else {
+                afd_events = self.poll_info.handles[0].events;
             }
+        }
 
-            let callback = unsafe { (*(status.overlapped() as *mut Overlapped)).callback };
+        afd_events &= self.user_evts;
 
-            trace!("select; -> got overlapped");
-            callback(status.entry());
+        if afd_events == 0 {
+            return None;
         }
 
-        trace!("returning");
-        Ok(())
-    }
+        // Codes to emulate ET in mio
```

Comment:
@PerfectLaugh 
Can you elaborate a bit on how this works?
In wepoll there's no way I can emulate edge-triggered, because I don't know when the user has attempted to read/write from a socket, so I don't know when to re-submit the AFD_POLL overlappped operation. I think in mio's case it might be possible because we can know when a socket read/write results in WSAEWOULDBLOCK. 
But I don't see how that works here.

Reply:
This work is in tcp.rs udp.rs. where I hooked all possible calls the WouldBlock could appear.

Reply:
And the codes here is to eliminate those events that have been read. So we can simulate proper ET behavior (probably?

Reply:
Could you explain how this code eliminates events that have been read?

Reply:
This code simply rely on assumption that mio can control and intercept every aspects that user will do to this socket. With this assumption (As long as rawsocket is not being manipulated by external library or raw api calls), we can wipe down the interests that user hasn't done it, simulating the ET behavior.